{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "43b86c5c-1904-4d89-9e8d-03d715adba18",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/software/cellgen/team274/lr26/miniforge3/envs/my-python/bin/python\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import glob\n",
    "import os\n",
    "from os import listdir\n",
    "from os.path import isfile, join\n",
    "import pysam \n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt # plots\n",
    "import scipy.sparse as sp\n",
    "from scipy.sparse import csr_matrix # matrices\n",
    "print(sys.executable)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "77d49546-1993-4be5-8f27-f2f7d5f8753b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "base_dir = \"/lustre/scratch126/casm/team274sb/lr26/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "bf3f0e3a-389a-4b43-919b-07711743ecfe",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>id</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>qual</th>\n",
       "      <th>filter</th>\n",
       "      <th>info</th>\n",
       "      <th>format</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10039</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>ACC</td>\n",
       "      <td>7</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CSQ=CC|upstream_gene_variant|MODIFIER|DDX11L1|...</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:C</td>\n",
       "      <td>0/1:4:5:2:0.4:DV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10050</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>AAC</td>\n",
       "      <td>6.5</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CSQ=AC|upstream_gene_variant|MODIFIER|DDX11L1|...</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:C</td>\n",
       "      <td>0/1:4:6:2:0.333333:DV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10093</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>ACC</td>\n",
       "      <td>6.7</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CSQ=CC|upstream_gene_variant|MODIFIER|DDX11L1|...</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:C</td>\n",
       "      <td>1/1:3:6:2:0.333333:DV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10105</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.1</td>\n",
       "      <td>refCall</td>\n",
       "      <td>CSQ=G|upstream_gene_variant|MODIFIER|DDX11L1|E...</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:C</td>\n",
       "      <td>./.:18:6:2:0.333333:DV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10108</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>11.9</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CSQ=T|upstream_gene_variant|MODIFIER|DDX11L1|E...</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:C</td>\n",
       "      <td>1/1:8:6:5:0.833333:DV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6986154</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56887636</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>24</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CSQ=G|intergenic_variant|MODIFIER|||||||||||||...</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:C</td>\n",
       "      <td>1/1:24:26:26:1:P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6986155</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56887659</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>23</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CSQ=G|intergenic_variant|MODIFIER|||||||||||||...</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:C</td>\n",
       "      <td>1/1:23:26:26:1:P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6986156</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56887837</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>17</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CSQ=A|intergenic_variant|MODIFIER|||||||||||||...</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:C</td>\n",
       "      <td>1/1:17:26:26:1:P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6986157</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56887844</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>21</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CSQ=C|intergenic_variant|MODIFIER|||||||||||||...</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:C</td>\n",
       "      <td>1/1:21:26:26:1:P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6986158</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56887853</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.2</td>\n",
       "      <td>refCall</td>\n",
       "      <td>CSQ=T|intergenic_variant|MODIFIER|||||||||||||...</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:C</td>\n",
       "      <td>./.:14:26:4:0.153846:DV</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6986159 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr       pos id ref  alt  qual   filter  \\\n",
       "0        chr1     10039  .   A  ACC     7     PASS   \n",
       "1        chr1     10050  .   A  AAC   6.5     PASS   \n",
       "2        chr1     10093  .   A  ACC   6.7     PASS   \n",
       "3        chr1     10105  .   A    G   0.1  refCall   \n",
       "4        chr1     10108  .   C   CT  11.9     PASS   \n",
       "...       ...       ... ..  ..  ...   ...      ...   \n",
       "6986154  chrY  56887636  .   T    G    24     PASS   \n",
       "6986155  chrY  56887659  .   T    G    23     PASS   \n",
       "6986156  chrY  56887837  .   G    A    17     PASS   \n",
       "6986157  chrY  56887844  .   T    C    21     PASS   \n",
       "6986158  chrY  56887853  .   G    T   0.2  refCall   \n",
       "\n",
       "                                                      info             format  \\\n",
       "0        CSQ=CC|upstream_gene_variant|MODIFIER|DDX11L1|...  GT:GQ:DP:AD:VAF:C   \n",
       "1        CSQ=AC|upstream_gene_variant|MODIFIER|DDX11L1|...  GT:GQ:DP:AD:VAF:C   \n",
       "2        CSQ=CC|upstream_gene_variant|MODIFIER|DDX11L1|...  GT:GQ:DP:AD:VAF:C   \n",
       "3        CSQ=G|upstream_gene_variant|MODIFIER|DDX11L1|E...  GT:GQ:DP:AD:VAF:C   \n",
       "4        CSQ=T|upstream_gene_variant|MODIFIER|DDX11L1|E...  GT:GQ:DP:AD:VAF:C   \n",
       "...                                                    ...                ...   \n",
       "6986154  CSQ=G|intergenic_variant|MODIFIER|||||||||||||...  GT:GQ:DP:AD:VAF:C   \n",
       "6986155  CSQ=G|intergenic_variant|MODIFIER|||||||||||||...  GT:GQ:DP:AD:VAF:C   \n",
       "6986156  CSQ=A|intergenic_variant|MODIFIER|||||||||||||...  GT:GQ:DP:AD:VAF:C   \n",
       "6986157  CSQ=C|intergenic_variant|MODIFIER|||||||||||||...  GT:GQ:DP:AD:VAF:C   \n",
       "6986158  CSQ=T|intergenic_variant|MODIFIER|||||||||||||...  GT:GQ:DP:AD:VAF:C   \n",
       "\n",
       "                          sample  \n",
       "0               0/1:4:5:2:0.4:DV  \n",
       "1          0/1:4:6:2:0.333333:DV  \n",
       "2          1/1:3:6:2:0.333333:DV  \n",
       "3         ./.:18:6:2:0.333333:DV  \n",
       "4          1/1:8:6:5:0.833333:DV  \n",
       "...                          ...  \n",
       "6986154         1/1:24:26:26:1:P  \n",
       "6986155         1/1:23:26:26:1:P  \n",
       "6986156         1/1:17:26:26:1:P  \n",
       "6986157         1/1:21:26:26:1:P  \n",
       "6986158  ./.:14:26:4:0.153846:DV  \n",
       "\n",
       "[6986159 rows x 10 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(base_dir + \"pepper-tumor38/tumor_vep_annotated_cosmic_clinvar_dbsnp.vcf\",\n",
    "                 sep=\"\\t\",\n",
    "                 comment=\"#\",\n",
    "                 dtype=\"str\",\n",
    "                 header=None,\n",
    "                 names=[\"chr\",\"pos\",\"id\",\"ref\",\"alt\",\"qual\",\"filter\",\"info\",\"format\",\"sample\"])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "776778dc-170e-4f56-9c8c-999dfcffeb69",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def parse_sample_format(format_str, sample_str):\n",
    "    \"\"\"\n",
    "    Parses FORMAT and SAMPLE fields of a VCF into a dictionary\n",
    "    Handles numeric parsing and multi-allelic values robustly\n",
    "    \"\"\"\n",
    "    keys = format_str.split(\":\")\n",
    "    values = sample_str.split(\":\")\n",
    "    sample_dict = dict(zip(keys, values))\n",
    "\n",
    "    # Parse GT\n",
    "    sample_dict[\"GT\"] = sample_dict.get(\"GT\", \"\")\n",
    "\n",
    "    # Parse DP\n",
    "    try:\n",
    "        sample_dict[\"DP\"] = int(sample_dict.get(\"DP\", 0))\n",
    "    except ValueError:\n",
    "        sample_dict[\"DP\"] = 0\n",
    "\n",
    "    # Parse AD\n",
    "    ad_val = sample_dict.get(\"AD\", \"0\")\n",
    "    if \",\" in ad_val:\n",
    "        try:\n",
    "            ad_list = [int(a) for a in ad_val.split(\",\") if a.isdigit()]\n",
    "            sample_dict[\"AD\"] = sum(ad_list[1:])  # Only ALT allele depth(s)\n",
    "        except:\n",
    "            sample_dict[\"AD\"] = 0\n",
    "    else:\n",
    "        sample_dict[\"AD\"] = int(ad_val) if ad_val.isdigit() else 0\n",
    "\n",
    "    # Parse VAF (handle multiple values by taking max)\n",
    "    vaf_val = sample_dict.get(\"VAF\", \"0.0\")\n",
    "    try:\n",
    "        if \",\" in vaf_val:\n",
    "            vaf_list = [float(v) for v in vaf_val.split(\",\") if v]\n",
    "            sample_dict[\"VAF\"] = max(vaf_list) if vaf_list else 0.0\n",
    "        else:\n",
    "            sample_dict[\"VAF\"] = float(vaf_val)\n",
    "    except:\n",
    "        sample_dict[\"VAF\"] = 0.0\n",
    "\n",
    "    return sample_dict\n",
    "\n",
    "\n",
    "def get_genotype_from_GT(gt_str):\n",
    "    \"\"\"\n",
    "    Determines genotype type based on GT field\n",
    "    \"\"\"\n",
    "    alleles = gt_str.replace('|', '/').split('/')\n",
    "    if len(alleles) == 2:\n",
    "        if alleles[0] == alleles[1]:\n",
    "            return \"homozygous\"\n",
    "        else:\n",
    "            return \"heterozygous\"\n",
    "    return \"unknown\"\n",
    "\n",
    "def extract_gt(sample_str, format_str):\n",
    "    keys = format_str.split(\":\")\n",
    "    values = sample_str.split(\":\")\n",
    "    format_dict = dict(zip(keys, values))\n",
    "    gt = format_dict.get(\"GT\", \"\")\n",
    "    return get_genotype_from_GT(gt)\n",
    "\n",
    "def filter_variant_by_sample(sample_dict, min_dp=10, min_ad=3, min_vaf=0.1):\n",
    "    \"\"\"\n",
    "    Apply basic quality filters on depth, allele depth, and VAF\n",
    "    \"\"\"\n",
    "    dp = sample_dict.get(\"DP\", 0)\n",
    "    ad = sample_dict.get(\"AD\", 0)\n",
    "    vaf = sample_dict.get(\"VAF\", 0.0)\n",
    "    \n",
    "    return (dp >= min_dp) and (ad >= min_ad) and (vaf >= min_vaf)\n",
    "\n",
    "def get_vep_annotations(info_str):\n",
    "    info_list = info_str.split(\";\")\n",
    "    info_dict = {}\n",
    "    csq_field_name = \"CSQ\"\n",
    "\n",
    "    csq_field_order = [\n",
    "        \"Allele\", \"Consequence\", \"IMPACT\", \"SYMBOL\", \"Gene\", \"Feature_type\",\n",
    "        \"Feature\", \"BIOTYPE\", \"EXON\", \"INTRON\", \"HGVSc\", \"HGVSp\", \"cDNA_position\",\n",
    "        \"CDS_position\", \"Protein_position\", \"Amino_acids\", \"Codons\",\n",
    "        \"Existing_variation\", \"DISTANCE\", \"STRAND\", \"FLAGS\", \"SYMBOL_SOURCE\",\n",
    "        \"HGNC_ID\", \"SOURCE\", \"dbSNP\", \"dbSNP_ID\", \"ClinVar\", \"ClinVar_CLINSIG\",\n",
    "        \"cosmic\", \"cosmic_COSMIC_ID\", \"gencode.sorted.gff3.gz\"\n",
    "    ]\n",
    "\n",
    "    annotations = []\n",
    "\n",
    "    for entry in info_list:\n",
    "        if \"=\" in entry:\n",
    "            key, value = entry.split(\"=\", 1)\n",
    "            if key == csq_field_name:\n",
    "                for ann in value.split(\",\"):\n",
    "                    fields = ann.split(\"|\")\n",
    "                    ann_dict = dict(zip(csq_field_order, fields))\n",
    "                    annotations.append(ann_dict)\n",
    "\n",
    "    return annotations\n",
    "\n",
    "def get_impact_list(ann_list):\n",
    "    impact_list = []\n",
    "    for annotation in ann_list:\n",
    "        impact = annotation[\"IMPACT\"]\n",
    "        impact_list.append(impact)\n",
    "    impact_list = list(set(impact_list))\n",
    "    return impact_list\n",
    "\n",
    "def get_impact_type(impact_list):\n",
    "    relevant = [\"MODERATE\",\"HIGH\"]\n",
    "    impact_type = \"\"\n",
    "    if (len(impact_list) != 0):\n",
    "        for impact in impact_list:\n",
    "            if (impact in(relevant)):\n",
    "                impact_type = \"relevant\"\n",
    "                break\n",
    "            else:\n",
    "                impact_type = \"non-relevant\"\n",
    "    else:\n",
    "        impact_type = \"relevant\"\n",
    "    return impact_type\n",
    "\n",
    "def get_clin_type(ann_list):\n",
    "    clin_sig_list = []\n",
    "    for annotation in ann_list:\n",
    "        clin_ann_list = annotation[\"ClinVar_CLINSIG\"].split(\"&\")\n",
    "        for clin_sig in clin_ann_list:\n",
    "            clin_ann_sub_list = clin_sig.split(\"/\")\n",
    "            for clin_sub_sig in clin_ann_sub_list:\n",
    "                clin_sig_list.append(clin_sub_sig)\n",
    "    clin_sig_list = set(clin_sig_list)\n",
    "    benign_list = [\"Benign\",\"Likely_benign\"]\n",
    "    benign_list = set(benign_list)\n",
    "    if (len(clin_sig_list.intersection(benign_list)) > 0):\n",
    "        clin_type = \"Benign\"\n",
    "    else:\n",
    "        clin_type = list(clin_sig_list)\n",
    "    return clin_type\n",
    "\n",
    "def extract_genes(annotations):\n",
    "    return list(set(\n",
    "        ann[\"SYMBOL\"] for ann in annotations if ann.get(\"SYMBOL\")\n",
    "    ))\n",
    "\n",
    "def extract_protein_positions(annotations):\n",
    "    return [\n",
    "        ann[\"Protein_position\"] for ann in annotations if ann.get(\"Protein_position\")\n",
    "    ]\n",
    "\n",
    "def extract_dbsnp_ids(annotations):\n",
    "    return list(set(\n",
    "        ann.get(\"dbSNP\") or ann.get(\"rsid\") for ann in annotations if ann.get(\"dbSNP\") or ann.get(\"rsid\")\n",
    "    ))\n",
    "\n",
    "def extract_consequences(annotations):\n",
    "    return list(set(\n",
    "        ann.get(\"Consequence\") or ann.get(\"Conseq\") for ann in annotations if ann.get(\"Consequence\") or ann.get(\"Conseq\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20c59cec-575a-422c-87f3-27aacaea5a76",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_sample_format(format_str, sample_str):\n",
    "    \"\"\"\n",
    "    Parses FORMAT and SAMPLE fields of a VCF into a dictionary\n",
    "    Handles numeric parsing and multi-allelic values robustly\n",
    "    \"\"\"\n",
    "    keys = format_str.split(\":\")\n",
    "    values = sample_str.split(\":\")\n",
    "    sample_dict = dict(zip(keys, values))\n",
    "\n",
    "    # Parse GT\n",
    "    sample_dict[\"GT\"] = sample_dict.get(\"GT\", \"\")\n",
    "\n",
    "    # Parse DP\n",
    "    try:\n",
    "        sample_dict[\"DP\"] = int(sample_dict.get(\"DP\", 0))\n",
    "    except ValueError:\n",
    "        sample_dict[\"DP\"] = 0\n",
    "\n",
    "    # Parse AD\n",
    "    ad_val = sample_dict.get(\"AD\", \"0\")\n",
    "    if \",\" in ad_val:\n",
    "        try:\n",
    "            ad_list = [int(a) for a in ad_val.split(\",\") if a.isdigit()]\n",
    "            sample_dict[\"AD\"] = sum(ad_list[1:])  # Only ALT allele depth(s)\n",
    "        except:\n",
    "            sample_dict[\"AD\"] = 0\n",
    "    else:\n",
    "        sample_dict[\"AD\"] = int(ad_val) if ad_val.isdigit() else 0\n",
    "\n",
    "    # Parse VAF (handle multiple values by taking max)\n",
    "    vaf_val = sample_dict.get(\"VAF\", \"0.0\")\n",
    "    try:\n",
    "        if \",\" in vaf_val:\n",
    "            vaf_list = [float(v) for v in vaf_val.split(\",\") if v]\n",
    "            sample_dict[\"VAF\"] = max(vaf_list) if vaf_list else 0.0\n",
    "        else:\n",
    "            sample_dict[\"VAF\"] = float(vaf_val)\n",
    "    except:\n",
    "        sample_dict[\"VAF\"] = 0.0\n",
    "\n",
    "    return sample_dict\n",
    "\n",
    "\n",
    "def get_genotype_from_GT(gt_str):\n",
    "    \"\"\"\n",
    "    Determines genotype type based on GT field\n",
    "    \"\"\"\n",
    "    alleles = gt_str.replace('|', '/').split('/')\n",
    "    if len(alleles) == 2:\n",
    "        if alleles[0] == alleles[1]:\n",
    "            return \"homozygous\"\n",
    "        else:\n",
    "            return \"heterozygous\"\n",
    "    return \"unknown\"\n",
    "\n",
    "def extract_gt(sample_str, format_str):\n",
    "    keys = format_str.split(\":\")\n",
    "    values = sample_str.split(\":\")\n",
    "    format_dict = dict(zip(keys, values))\n",
    "    gt = format_dict.get(\"GT\", \"\")\n",
    "    return get_genotype_from_GT(gt)\n",
    "\n",
    "def filter_variant_by_sample(sample_dict, min_dp=10, min_ad=3, min_vaf=0.1):\n",
    "    \"\"\"\n",
    "    Apply basic quality filters on depth, allele depth, and VAF\n",
    "    \"\"\"\n",
    "    dp = sample_dict.get(\"DP\", 0)\n",
    "    ad = sample_dict.get(\"AD\", 0)\n",
    "    vaf = sample_dict.get(\"VAF\", 0.0)\n",
    "    \n",
    "    return (dp >= min_dp) and (ad >= min_ad) and (vaf >= min_vaf)\n",
    "\n",
    "def get_vep_annotations(info_str):\n",
    "    info_list = info_str.split(\";\")\n",
    "    info_dict = {}\n",
    "    csq_field_name = \"CSQ\"\n",
    "\n",
    "    csq_field_order = [\n",
    "        \"Allele\", \"Consequence\", \"IMPACT\", \"SYMBOL\", \"Gene\", \"Feature_type\",\n",
    "        \"Feature\", \"BIOTYPE\", \"EXON\", \"INTRON\", \"HGVSc\", \"HGVSp\", \"cDNA_position\",\n",
    "        \"CDS_position\", \"Protein_position\", \"Amino_acids\", \"Codons\",\n",
    "        \"Existing_variation\", \"DISTANCE\", \"STRAND\", \"FLAGS\", \"SYMBOL_SOURCE\",\n",
    "        \"HGNC_ID\", \"SOURCE\", \"ClinVar\", \"ClinVar_ID\", \"dbSNP\", \"dbSNP_ID\",\n",
    "        \"chm13v2.0_RefSeq_Liftoff_v5.1.sorted.gff3.gz\"\n",
    "    ]\n",
    "\n",
    "    annotations = []\n",
    "\n",
    "    for entry in info_list:\n",
    "        if \"=\" in entry:\n",
    "            key, value = entry.split(\"=\", 1)\n",
    "            if key == csq_field_name:\n",
    "                for ann in value.split(\",\"):\n",
    "                    fields = ann.split(\"|\")\n",
    "                    ann_dict = dict(zip(csq_field_order, fields))\n",
    "                    annotations.append(ann_dict)\n",
    "\n",
    "    return annotations\n",
    "\n",
    "def get_impact_list(ann_list):\n",
    "    impact_list = []\n",
    "    for annotation in ann_list:\n",
    "        impact = annotation[\"IMPACT\"]\n",
    "        impact_list.append(impact)\n",
    "    impact_list = list(set(impact_list))\n",
    "    return impact_list\n",
    "\n",
    "def get_impact_type(impact_list):\n",
    "    relevant = [\"MODERATE\",\"HIGH\"]\n",
    "    impact_type = \"\"\n",
    "    if (len(impact_list) != 0):\n",
    "        for impact in impact_list:\n",
    "            if (impact in(relevant)):\n",
    "                impact_type = \"relevant\"\n",
    "                break\n",
    "            else:\n",
    "                impact_type = \"non-relevant\"\n",
    "    else:\n",
    "        impact_type = \"relevant\"\n",
    "    return impact_type\n",
    "\n",
    "def get_clin_type(ann_list):\n",
    "    clin_sig_list = []\n",
    "    for annotation in ann_list:\n",
    "        clin_ann_list = annotation[\"ClinVar_ID\"].split(\"&\")\n",
    "        for clin_sig in clin_ann_list:\n",
    "            clin_ann_sub_list = clin_sig.split(\"/\")\n",
    "            for clin_sub_sig in clin_ann_sub_list:\n",
    "                clin_sig_list.append(clin_sub_sig)\n",
    "    clin_sig_list = set(clin_sig_list)\n",
    "    benign_list = [\"Benign\",\"Likely_benign\"]\n",
    "    benign_list = set(benign_list)\n",
    "    if (len(clin_sig_list.intersection(benign_list)) > 0):\n",
    "        clin_type = \"Benign\"\n",
    "    else:\n",
    "        clin_type = list(clin_sig_list)\n",
    "    return clin_type\n",
    "\n",
    "def extract_genes(annotations):\n",
    "    return list(set(\n",
    "        ann[\"Gene\"] for ann in annotations if ann.get(\"Gene\")\n",
    "    ))\n",
    "\n",
    "def extract_protein_positions(annotations):\n",
    "    return [\n",
    "        ann[\"Protein_position\"] for ann in annotations if ann.get(\"Protein_position\")\n",
    "    ]\n",
    "\n",
    "def extract_dbsnp_ids(annotations):\n",
    "    return list(set(\n",
    "        ann.get(\"dbSNP\") or ann.get(\"rsid\") for ann in annotations if ann.get(\"dbSNP\") or ann.get(\"rsid\")\n",
    "    ))\n",
    "\n",
    "def extract_consequences(annotations):\n",
    "    return list(set(\n",
    "        ann.get(\"Consequence\") or ann.get(\"Conseq\") for ann in annotations if ann.get(\"Consequence\") or ann.get(\"Conseq\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5608bbfe-20d6-4df1-a8f3-4f8207ccf5e7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_filt = df.copy()\n",
    "df_filt[\"qual\"] = pd.to_numeric(df_filt[\"qual\"], errors='coerce')  # Convert in the copy\n",
    "\n",
    "df_filtered = df_filt[\n",
    "    (df_filt[\"filter\"] == \"PASS\") &\n",
    "    (df_filt[\"qual\"] > 5)\n",
    "]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2f91555b-274a-4f12-995e-94caa0c08f54",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>id</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>qual</th>\n",
       "      <th>filter</th>\n",
       "      <th>info</th>\n",
       "      <th>format</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10039</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>ACC</td>\n",
       "      <td>7.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CSQ=CC|upstream_gene_variant|MODIFIER|DDX11L1|...</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:C</td>\n",
       "      <td>0/1:4:5:2:0.4:DV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10050</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>AAC</td>\n",
       "      <td>6.5</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CSQ=AC|upstream_gene_variant|MODIFIER|DDX11L1|...</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:C</td>\n",
       "      <td>0/1:4:6:2:0.333333:DV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10093</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>ACC</td>\n",
       "      <td>6.7</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CSQ=CC|upstream_gene_variant|MODIFIER|DDX11L1|...</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:C</td>\n",
       "      <td>1/1:3:6:2:0.333333:DV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10108</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>11.9</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CSQ=T|upstream_gene_variant|MODIFIER|DDX11L1|E...</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:C</td>\n",
       "      <td>1/1:8:6:5:0.833333:DV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10217</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>8.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CSQ=C|upstream_gene_variant|MODIFIER|DDX11L1|E...</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:C</td>\n",
       "      <td>1/1:4:6:2:0.333333:DV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6986153</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56887631</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>26.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CSQ=A|intergenic_variant|MODIFIER|||||||||||||...</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:C</td>\n",
       "      <td>1/1:26:26:26:1:P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6986154</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56887636</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>24.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CSQ=G|intergenic_variant|MODIFIER|||||||||||||...</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:C</td>\n",
       "      <td>1/1:24:26:26:1:P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6986155</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56887659</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>23.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CSQ=G|intergenic_variant|MODIFIER|||||||||||||...</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:C</td>\n",
       "      <td>1/1:23:26:26:1:P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6986156</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56887837</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>17.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CSQ=A|intergenic_variant|MODIFIER|||||||||||||...</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:C</td>\n",
       "      <td>1/1:17:26:26:1:P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6986157</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56887844</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>21.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CSQ=C|intergenic_variant|MODIFIER|||||||||||||...</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:C</td>\n",
       "      <td>1/1:21:26:26:1:P</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4898779 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr       pos id ref  alt  qual filter  \\\n",
       "0        chr1     10039  .   A  ACC   7.0   PASS   \n",
       "1        chr1     10050  .   A  AAC   6.5   PASS   \n",
       "2        chr1     10093  .   A  ACC   6.7   PASS   \n",
       "4        chr1     10108  .   C   CT  11.9   PASS   \n",
       "14       chr1     10217  .   A    C   8.0   PASS   \n",
       "...       ...       ... ..  ..  ...   ...    ...   \n",
       "6986153  chrY  56887631  .   C    A  26.0   PASS   \n",
       "6986154  chrY  56887636  .   T    G  24.0   PASS   \n",
       "6986155  chrY  56887659  .   T    G  23.0   PASS   \n",
       "6986156  chrY  56887837  .   G    A  17.0   PASS   \n",
       "6986157  chrY  56887844  .   T    C  21.0   PASS   \n",
       "\n",
       "                                                      info             format  \\\n",
       "0        CSQ=CC|upstream_gene_variant|MODIFIER|DDX11L1|...  GT:GQ:DP:AD:VAF:C   \n",
       "1        CSQ=AC|upstream_gene_variant|MODIFIER|DDX11L1|...  GT:GQ:DP:AD:VAF:C   \n",
       "2        CSQ=CC|upstream_gene_variant|MODIFIER|DDX11L1|...  GT:GQ:DP:AD:VAF:C   \n",
       "4        CSQ=T|upstream_gene_variant|MODIFIER|DDX11L1|E...  GT:GQ:DP:AD:VAF:C   \n",
       "14       CSQ=C|upstream_gene_variant|MODIFIER|DDX11L1|E...  GT:GQ:DP:AD:VAF:C   \n",
       "...                                                    ...                ...   \n",
       "6986153  CSQ=A|intergenic_variant|MODIFIER|||||||||||||...  GT:GQ:DP:AD:VAF:C   \n",
       "6986154  CSQ=G|intergenic_variant|MODIFIER|||||||||||||...  GT:GQ:DP:AD:VAF:C   \n",
       "6986155  CSQ=G|intergenic_variant|MODIFIER|||||||||||||...  GT:GQ:DP:AD:VAF:C   \n",
       "6986156  CSQ=A|intergenic_variant|MODIFIER|||||||||||||...  GT:GQ:DP:AD:VAF:C   \n",
       "6986157  CSQ=C|intergenic_variant|MODIFIER|||||||||||||...  GT:GQ:DP:AD:VAF:C   \n",
       "\n",
       "                        sample  \n",
       "0             0/1:4:5:2:0.4:DV  \n",
       "1        0/1:4:6:2:0.333333:DV  \n",
       "2        1/1:3:6:2:0.333333:DV  \n",
       "4        1/1:8:6:5:0.833333:DV  \n",
       "14       1/1:4:6:2:0.333333:DV  \n",
       "...                        ...  \n",
       "6986153       1/1:26:26:26:1:P  \n",
       "6986154       1/1:24:26:26:1:P  \n",
       "6986155       1/1:23:26:26:1:P  \n",
       "6986156       1/1:17:26:26:1:P  \n",
       "6986157       1/1:21:26:26:1:P  \n",
       "\n",
       "[4898779 rows x 10 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3394a9e2-424d-4945-9f9a-22c402048b6a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2469804/3858194818.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_filtered[\"sample_info\"] = df_filtered.apply(lambda row: parse_sample_format(row[\"format\"], row[\"sample\"]), axis=1)\n",
      "/tmp/ipykernel_2469804/3858194818.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_filtered[\"DP\"]  = df_filtered[\"sample_info\"].apply(lambda x: x[\"DP\"])\n",
      "/tmp/ipykernel_2469804/3858194818.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_filtered[\"AD\"]  = df_filtered[\"sample_info\"].apply(lambda x: x[\"AD\"])\n",
      "/tmp/ipykernel_2469804/3858194818.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_filtered[\"VAF\"] = df_filtered[\"sample_info\"].apply(lambda x: x[\"VAF\"])\n",
      "/tmp/ipykernel_2469804/3858194818.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_filtered[\"GT\"]  = df_filtered[\"sample_info\"].apply(lambda x: x[\"GT\"])\n",
      "/tmp/ipykernel_2469804/3858194818.py:9: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_filtered[\"genotype\"] = df_filtered[\"GT\"].apply(get_genotype_from_GT)\n",
      "/tmp/ipykernel_2469804/3858194818.py:12: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_filtered[\"pass_sample_filters\"] = df_filtered[\"sample_info\"].apply(filter_variant_by_sample)\n",
      "/tmp/ipykernel_2469804/3858194818.py:15: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_filtered[\"vep_ann\"] = df_filtered[\"info\"].apply(get_vep_annotations)\n",
      "/tmp/ipykernel_2469804/3858194818.py:18: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_filtered[\"impact_list\"] = df_filtered[\"vep_ann\"].apply(get_impact_list)\n",
      "/tmp/ipykernel_2469804/3858194818.py:19: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_filtered[\"impact_type\"] = df_filtered[\"impact_list\"].apply(get_impact_type)\n",
      "/tmp/ipykernel_2469804/3858194818.py:22: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_filtered[\"clin_sig\"] = df_filtered[\"vep_ann\"].apply(get_clin_type)\n",
      "/tmp/ipykernel_2469804/3858194818.py:24: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_filtered[\"genes\"] = df_filtered[\"vep_ann\"].apply(extract_genes)\n",
      "/tmp/ipykernel_2469804/3858194818.py:25: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_filtered[\"protein_positions\"] = df_filtered[\"vep_ann\"].apply(extract_protein_positions)\n",
      "/tmp/ipykernel_2469804/3858194818.py:26: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_filtered[\"dbsnp_ids\"] = df_filtered[\"vep_ann\"].apply(extract_dbsnp_ids)\n",
      "/tmp/ipykernel_2469804/3858194818.py:27: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_filtered[\"consequences\"] = df_filtered[\"vep_ann\"].apply(extract_consequences)\n"
     ]
    }
   ],
   "source": [
    "# Extract sample info\n",
    "df_filtered[\"sample_info\"] = df_filtered.apply(lambda row: parse_sample_format(row[\"format\"], row[\"sample\"]), axis=1)\n",
    "\n",
    "# Add specific fields as separate columns\n",
    "df_filtered[\"DP\"]  = df_filtered[\"sample_info\"].apply(lambda x: x[\"DP\"])\n",
    "df_filtered[\"AD\"]  = df_filtered[\"sample_info\"].apply(lambda x: x[\"AD\"])\n",
    "df_filtered[\"VAF\"] = df_filtered[\"sample_info\"].apply(lambda x: x[\"VAF\"])\n",
    "df_filtered[\"GT\"]  = df_filtered[\"sample_info\"].apply(lambda x: x[\"GT\"])\n",
    "df_filtered[\"genotype\"] = df_filtered[\"GT\"].apply(get_genotype_from_GT)\n",
    "\n",
    "# Apply filtering\n",
    "df_filtered[\"pass_sample_filters\"] = df_filtered[\"sample_info\"].apply(filter_variant_by_sample)\n",
    "\n",
    "# Extract VEP annotations\n",
    "df_filtered[\"vep_ann\"] = df_filtered[\"info\"].apply(get_vep_annotations)\n",
    "\n",
    "# Impact info\n",
    "df_filtered[\"impact_list\"] = df_filtered[\"vep_ann\"].apply(get_impact_list)\n",
    "df_filtered[\"impact_type\"] = df_filtered[\"impact_list\"].apply(get_impact_type)\n",
    "\n",
    "# Clinical significance\n",
    "df_filtered[\"clin_sig\"] = df_filtered[\"vep_ann\"].apply(get_clin_type)\n",
    "\n",
    "df_filtered[\"genes\"] = df_filtered[\"vep_ann\"].apply(extract_genes)\n",
    "df_filtered[\"protein_positions\"] = df_filtered[\"vep_ann\"].apply(extract_protein_positions)\n",
    "df_filtered[\"dbsnp_ids\"] = df_filtered[\"vep_ann\"].apply(extract_dbsnp_ids)\n",
    "df_filtered[\"consequences\"] = df_filtered[\"vep_ann\"].apply(extract_consequences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "895503e3-84ef-4c28-9668-e4178a0dd5a2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>id</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>qual</th>\n",
       "      <th>filter</th>\n",
       "      <th>info</th>\n",
       "      <th>format</th>\n",
       "      <th>sample</th>\n",
       "      <th>...</th>\n",
       "      <th>genotype</th>\n",
       "      <th>pass_sample_filters</th>\n",
       "      <th>vep_ann</th>\n",
       "      <th>impact_list</th>\n",
       "      <th>impact_type</th>\n",
       "      <th>clin_sig</th>\n",
       "      <th>genes</th>\n",
       "      <th>protein_positions</th>\n",
       "      <th>dbsnp_ids</th>\n",
       "      <th>consequences</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10039</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>ACC</td>\n",
       "      <td>7.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CSQ=CC|upstream_gene_variant|MODIFIER|DDX11L1|...</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:C</td>\n",
       "      <td>0/1:4:5:2:0.4:DV</td>\n",
       "      <td>...</td>\n",
       "      <td>heterozygous</td>\n",
       "      <td>False</td>\n",
       "      <td>[{'Allele': 'CC', 'Consequence': 'upstream_gen...</td>\n",
       "      <td>[MODIFIER]</td>\n",
       "      <td>non-relevant</td>\n",
       "      <td>[]</td>\n",
       "      <td>[DDX11L2, DDX11L1, WASH7P]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[upstream_gene_variant, downstream_gene_variant]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10050</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>AAC</td>\n",
       "      <td>6.5</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CSQ=AC|upstream_gene_variant|MODIFIER|DDX11L1|...</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:C</td>\n",
       "      <td>0/1:4:6:2:0.333333:DV</td>\n",
       "      <td>...</td>\n",
       "      <td>heterozygous</td>\n",
       "      <td>False</td>\n",
       "      <td>[{'Allele': 'AC', 'Consequence': 'upstream_gen...</td>\n",
       "      <td>[MODIFIER]</td>\n",
       "      <td>non-relevant</td>\n",
       "      <td>[]</td>\n",
       "      <td>[DDX11L2, DDX11L1, WASH7P]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[upstream_gene_variant, downstream_gene_variant]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10093</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>ACC</td>\n",
       "      <td>6.7</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CSQ=CC|upstream_gene_variant|MODIFIER|DDX11L1|...</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:C</td>\n",
       "      <td>1/1:3:6:2:0.333333:DV</td>\n",
       "      <td>...</td>\n",
       "      <td>homozygous</td>\n",
       "      <td>False</td>\n",
       "      <td>[{'Allele': 'CC', 'Consequence': 'upstream_gen...</td>\n",
       "      <td>[MODIFIER]</td>\n",
       "      <td>non-relevant</td>\n",
       "      <td>[]</td>\n",
       "      <td>[DDX11L2, DDX11L1, WASH7P]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[upstream_gene_variant, downstream_gene_variant]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10108</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>11.9</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CSQ=T|upstream_gene_variant|MODIFIER|DDX11L1|E...</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:C</td>\n",
       "      <td>1/1:8:6:5:0.833333:DV</td>\n",
       "      <td>...</td>\n",
       "      <td>homozygous</td>\n",
       "      <td>False</td>\n",
       "      <td>[{'Allele': 'T', 'Consequence': 'upstream_gene...</td>\n",
       "      <td>[MODIFIER]</td>\n",
       "      <td>non-relevant</td>\n",
       "      <td>[]</td>\n",
       "      <td>[DDX11L2, DDX11L1, WASH7P]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[upstream_gene_variant, downstream_gene_variant]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10217</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>8.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CSQ=C|upstream_gene_variant|MODIFIER|DDX11L1|E...</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:C</td>\n",
       "      <td>1/1:4:6:2:0.333333:DV</td>\n",
       "      <td>...</td>\n",
       "      <td>homozygous</td>\n",
       "      <td>False</td>\n",
       "      <td>[{'Allele': 'C', 'Consequence': 'upstream_gene...</td>\n",
       "      <td>[MODIFIER]</td>\n",
       "      <td>non-relevant</td>\n",
       "      <td>[]</td>\n",
       "      <td>[DDX11L2, DDX11L1, WASH7P]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[upstream_gene_variant, downstream_gene_variant]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6986153</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56887631</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>26.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CSQ=A|intergenic_variant|MODIFIER|||||||||||||...</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:C</td>\n",
       "      <td>1/1:26:26:26:1:P</td>\n",
       "      <td>...</td>\n",
       "      <td>homozygous</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'Allele': 'A', 'Consequence': 'intergenic_va...</td>\n",
       "      <td>[MODIFIER]</td>\n",
       "      <td>non-relevant</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[intergenic_variant]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6986154</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56887636</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>24.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CSQ=G|intergenic_variant|MODIFIER|||||||||||||...</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:C</td>\n",
       "      <td>1/1:24:26:26:1:P</td>\n",
       "      <td>...</td>\n",
       "      <td>homozygous</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'Allele': 'G', 'Consequence': 'intergenic_va...</td>\n",
       "      <td>[MODIFIER]</td>\n",
       "      <td>non-relevant</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[rs111327212]</td>\n",
       "      <td>[intergenic_variant]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6986155</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56887659</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>23.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CSQ=G|intergenic_variant|MODIFIER|||||||||||||...</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:C</td>\n",
       "      <td>1/1:23:26:26:1:P</td>\n",
       "      <td>...</td>\n",
       "      <td>homozygous</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'Allele': 'G', 'Consequence': 'intergenic_va...</td>\n",
       "      <td>[MODIFIER]</td>\n",
       "      <td>non-relevant</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[rs376828276]</td>\n",
       "      <td>[intergenic_variant]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6986156</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56887837</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>17.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CSQ=A|intergenic_variant|MODIFIER|||||||||||||...</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:C</td>\n",
       "      <td>1/1:17:26:26:1:P</td>\n",
       "      <td>...</td>\n",
       "      <td>homozygous</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'Allele': 'A', 'Consequence': 'intergenic_va...</td>\n",
       "      <td>[MODIFIER]</td>\n",
       "      <td>non-relevant</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[intergenic_variant]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6986157</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56887844</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>21.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CSQ=C|intergenic_variant|MODIFIER|||||||||||||...</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:C</td>\n",
       "      <td>1/1:21:26:26:1:P</td>\n",
       "      <td>...</td>\n",
       "      <td>homozygous</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'Allele': 'C', 'Consequence': 'intergenic_va...</td>\n",
       "      <td>[MODIFIER]</td>\n",
       "      <td>non-relevant</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[intergenic_variant]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4898779 rows Ã— 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr       pos id ref  alt  qual filter  \\\n",
       "0        chr1     10039  .   A  ACC   7.0   PASS   \n",
       "1        chr1     10050  .   A  AAC   6.5   PASS   \n",
       "2        chr1     10093  .   A  ACC   6.7   PASS   \n",
       "4        chr1     10108  .   C   CT  11.9   PASS   \n",
       "14       chr1     10217  .   A    C   8.0   PASS   \n",
       "...       ...       ... ..  ..  ...   ...    ...   \n",
       "6986153  chrY  56887631  .   C    A  26.0   PASS   \n",
       "6986154  chrY  56887636  .   T    G  24.0   PASS   \n",
       "6986155  chrY  56887659  .   T    G  23.0   PASS   \n",
       "6986156  chrY  56887837  .   G    A  17.0   PASS   \n",
       "6986157  chrY  56887844  .   T    C  21.0   PASS   \n",
       "\n",
       "                                                      info             format  \\\n",
       "0        CSQ=CC|upstream_gene_variant|MODIFIER|DDX11L1|...  GT:GQ:DP:AD:VAF:C   \n",
       "1        CSQ=AC|upstream_gene_variant|MODIFIER|DDX11L1|...  GT:GQ:DP:AD:VAF:C   \n",
       "2        CSQ=CC|upstream_gene_variant|MODIFIER|DDX11L1|...  GT:GQ:DP:AD:VAF:C   \n",
       "4        CSQ=T|upstream_gene_variant|MODIFIER|DDX11L1|E...  GT:GQ:DP:AD:VAF:C   \n",
       "14       CSQ=C|upstream_gene_variant|MODIFIER|DDX11L1|E...  GT:GQ:DP:AD:VAF:C   \n",
       "...                                                    ...                ...   \n",
       "6986153  CSQ=A|intergenic_variant|MODIFIER|||||||||||||...  GT:GQ:DP:AD:VAF:C   \n",
       "6986154  CSQ=G|intergenic_variant|MODIFIER|||||||||||||...  GT:GQ:DP:AD:VAF:C   \n",
       "6986155  CSQ=G|intergenic_variant|MODIFIER|||||||||||||...  GT:GQ:DP:AD:VAF:C   \n",
       "6986156  CSQ=A|intergenic_variant|MODIFIER|||||||||||||...  GT:GQ:DP:AD:VAF:C   \n",
       "6986157  CSQ=C|intergenic_variant|MODIFIER|||||||||||||...  GT:GQ:DP:AD:VAF:C   \n",
       "\n",
       "                        sample  ...      genotype  pass_sample_filters  \\\n",
       "0             0/1:4:5:2:0.4:DV  ...  heterozygous                False   \n",
       "1        0/1:4:6:2:0.333333:DV  ...  heterozygous                False   \n",
       "2        1/1:3:6:2:0.333333:DV  ...    homozygous                False   \n",
       "4        1/1:8:6:5:0.833333:DV  ...    homozygous                False   \n",
       "14       1/1:4:6:2:0.333333:DV  ...    homozygous                False   \n",
       "...                        ...  ...           ...                  ...   \n",
       "6986153       1/1:26:26:26:1:P  ...    homozygous                 True   \n",
       "6986154       1/1:24:26:26:1:P  ...    homozygous                 True   \n",
       "6986155       1/1:23:26:26:1:P  ...    homozygous                 True   \n",
       "6986156       1/1:17:26:26:1:P  ...    homozygous                 True   \n",
       "6986157       1/1:21:26:26:1:P  ...    homozygous                 True   \n",
       "\n",
       "                                                   vep_ann  impact_list  \\\n",
       "0        [{'Allele': 'CC', 'Consequence': 'upstream_gen...   [MODIFIER]   \n",
       "1        [{'Allele': 'AC', 'Consequence': 'upstream_gen...   [MODIFIER]   \n",
       "2        [{'Allele': 'CC', 'Consequence': 'upstream_gen...   [MODIFIER]   \n",
       "4        [{'Allele': 'T', 'Consequence': 'upstream_gene...   [MODIFIER]   \n",
       "14       [{'Allele': 'C', 'Consequence': 'upstream_gene...   [MODIFIER]   \n",
       "...                                                    ...          ...   \n",
       "6986153  [{'Allele': 'A', 'Consequence': 'intergenic_va...   [MODIFIER]   \n",
       "6986154  [{'Allele': 'G', 'Consequence': 'intergenic_va...   [MODIFIER]   \n",
       "6986155  [{'Allele': 'G', 'Consequence': 'intergenic_va...   [MODIFIER]   \n",
       "6986156  [{'Allele': 'A', 'Consequence': 'intergenic_va...   [MODIFIER]   \n",
       "6986157  [{'Allele': 'C', 'Consequence': 'intergenic_va...   [MODIFIER]   \n",
       "\n",
       "          impact_type clin_sig                       genes protein_positions  \\\n",
       "0        non-relevant       []  [DDX11L2, DDX11L1, WASH7P]                []   \n",
       "1        non-relevant       []  [DDX11L2, DDX11L1, WASH7P]                []   \n",
       "2        non-relevant       []  [DDX11L2, DDX11L1, WASH7P]                []   \n",
       "4        non-relevant       []  [DDX11L2, DDX11L1, WASH7P]                []   \n",
       "14       non-relevant       []  [DDX11L2, DDX11L1, WASH7P]                []   \n",
       "...               ...      ...                         ...               ...   \n",
       "6986153  non-relevant       []                          []                []   \n",
       "6986154  non-relevant       []                          []                []   \n",
       "6986155  non-relevant       []                          []                []   \n",
       "6986156  non-relevant       []                          []                []   \n",
       "6986157  non-relevant       []                          []                []   \n",
       "\n",
       "             dbsnp_ids                                      consequences  \n",
       "0                   []  [upstream_gene_variant, downstream_gene_variant]  \n",
       "1                   []  [upstream_gene_variant, downstream_gene_variant]  \n",
       "2                   []  [upstream_gene_variant, downstream_gene_variant]  \n",
       "4                   []  [upstream_gene_variant, downstream_gene_variant]  \n",
       "14                  []  [upstream_gene_variant, downstream_gene_variant]  \n",
       "...                ...                                               ...  \n",
       "6986153             []                              [intergenic_variant]  \n",
       "6986154  [rs111327212]                              [intergenic_variant]  \n",
       "6986155  [rs376828276]                              [intergenic_variant]  \n",
       "6986156             []                              [intergenic_variant]  \n",
       "6986157             []                              [intergenic_variant]  \n",
       "\n",
       "[4898779 rows x 25 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d875cea4-e270-42f6-af75-eaeda02fbfec",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_filtered.to_csv(base_dir + \"pepper-blood/blood_vep_annotated_filtered_new.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a2fc2892-712f-4088-ac40-6547b52ceabf",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_fin = df_filtered[\n",
    "    (df_filtered[\"filter\"] == \"PASS\") &\n",
    "    (df_filtered[\"qual\"] > 5) &\n",
    "    (df_filtered[\"AD\"] > 5) &\n",
    "    (df_filtered[\"impact_type\"] == \"relevant\") &\n",
    "    (df_filtered[\"clin_sig\"].apply(lambda x: any(s.strip() for s in x if isinstance(s, str))))\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59c49f24-f5d4-4678-8f35-a80f8f72cb66",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_fin"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c14ba754-58d9-4a08-85b6-d81ff20e7eb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_fin.to_csv(base_dir + \"pepper-tumor38/tumor_all_clinvar_dbsnp_filtered_hg38.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e01b5eab-31f0-483a-ae23-6532796be18c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "281be266-504d-4b9b-a9ae-59bf3b1d39d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the PanelApp gene panel file\n",
    "panel_df = pd.read_csv(\"/nfs/users/nfs_l/lr26/shells/panelapp_childhood_solid_tumors_420.tsv\", sep=\"\\t\")\n",
    "\n",
    "# Extract unique gene names from the first column (assuming itâ€™s called \"Entity Name\")\n",
    "cancer_genes = panel_df[\"Entity Name\"].dropna().unique().tolist()\n",
    "cancer_genes\n",
    "\n",
    "def get_affected_genes(ann_list):\n",
    "    gene_list = []\n",
    "    for annotation in ann_list:\n",
    "        symbol = annotation[\"SYMBOL\"]\n",
    "        gene_list.append(symbol)\n",
    "    gene_list = list(set(gene_list))\n",
    "    return gene_list\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c33d116b-5ebd-4368-9f7a-5bf6f5a91d06",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_df              = df_filtered.copy()\n",
    "gene_df[\"gene_list\"] = gene_df[\"vep_ann\"].apply(lambda x: get_affected_genes(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90c1bb7b-0af9-4c28-ae4e-42555c65015d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_cancer = gene_df[gene_df[\"genes\"].apply(lambda gene_list: any(gene in cancer_genes for gene in gene_list))]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80feb635-abe4-4f43-8eaa-57f902162978",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1445d23-f823-4142-a6e0-ed807c54c4b9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_cancer.to_csv(base_dir + \"pepper-tumor1B01/tumor_vep_annotated_filtered_final_new_cancer_panel.tsv\", sep=\"\\t\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "My Py",
   "language": "python",
   "name": "my-python"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
