{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "81f719d3-e551-46ad-9e73-f10b542910ed",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Output saved to /lustre/scratch126/casm/team274sb/lr26/hg38/cosmic_summaryhg38.tsv\n"
     ]
    }
   ],
   "source": [
    "import pysam\n",
    "from collections import defaultdict\n",
    "\n",
    "# Input and output files\n",
    "vcf_file = '/lustre/scratch126/casm/team274sb/lr26/hg38/cleaned_cosmic_hg38.vcf.gz'\n",
    "output_file = '/lustre/scratch126/casm/team274sb/lr26/hg38/cosmic_summaryhg38.tsv'\n",
    "\n",
    "# Dictionary to store variant data and counts\n",
    "variant_data = defaultdict(lambda: {\n",
    "    'gene': '.',\n",
    "    'aa': '.',\n",
    "    'desc': '.',\n",
    "    'count': 0\n",
    "})\n",
    "\n",
    "# Open VCF file\n",
    "vcf = pysam.VariantFile(vcf_file, \"r\")\n",
    "\n",
    "# Iterate through each record in the VCF\n",
    "for record in vcf.fetch():\n",
    "    # Define unique variant key by genomic coordinates and alleles\n",
    "    key = (record.chrom, record.pos, record.ref, tuple(record.alts))\n",
    "\n",
    "    # Update count and store annotation if it's the first time\n",
    "    variant_data[key]['count'] += 1\n",
    "\n",
    "    # Store annotations (just use the first one encountered)\n",
    "    if variant_data[key]['gene'] == '.':\n",
    "        variant_data[key]['gene'] = record.info.get('GENE_SYMBOL', '.')\n",
    "        variant_data[key]['aa'] = record.info.get('MUTATION_AA', '.')\n",
    "        variant_data[key]['desc'] = record.info.get('MUTATION_DESCRIPTION', '.')\n",
    "\n",
    "# Write results to a TSV file\n",
    "with open(output_file, 'w') as out:\n",
    "    out.write(\"CHROM\\tPOS\\tREF\\tALT\\tCOUNT\\tGENE_SYMBOL\\tMUTATION_AA\\tMUTATION_DESCRIPTION\\n\")\n",
    "    for (chrom, pos, ref, alts), info in variant_data.items():\n",
    "        alt_str = \",\".join(alts)\n",
    "        out.write(f\"{chrom}\\t{pos}\\t{ref}\\t{alt_str}\\t{info['count']}\\t{info['gene']}\\t{info['aa']}\\t{info['desc']}\\n\")\n",
    "\n",
    "print(\"✅ Output saved to\", output_file)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "72220249-77b1-4394-ac6e-c6b0d3e4fb7b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Done. Output saved to: /lustre/scratch126/casm/team274sb/lr26/T2T/cosmic_summaryt2t.tsv\n"
     ]
    }
   ],
   "source": [
    "# Input and output files\n",
    "vcf_file = '/lustre/scratch126/casm/team274sb/lr26/T2T/Cosmic_MutantCensus_sorted_T2T.vcf.gz'\n",
    "output_file = '/lustre/scratch126/casm/team274sb/lr26/T2T/cosmic_summaryt2t.tsv'\n",
    "\n",
    "# Dictionary to store variant data and counts\n",
    "variant_data = defaultdict(lambda: {\n",
    "    'gene': '.',\n",
    "    'aa': '.',\n",
    "    'desc': '.',\n",
    "    'count': 0\n",
    "})\n",
    "\n",
    "# Open the VCF file\n",
    "vcf = pysam.VariantFile(vcf_file, \"r\")\n",
    "\n",
    "# Iterate over records\n",
    "for record in vcf.fetch():\n",
    "    if record.alts is None:\n",
    "        continue  # Skip invalid/missing ALT entries\n",
    "\n",
    "    for alt in record.alts:\n",
    "        # ⚠️ Make a precise key: include ALT separately\n",
    "        key = (record.chrom, record.pos, record.ref, alt)\n",
    "\n",
    "        # Increment count\n",
    "        variant_data[key]['count'] += 1\n",
    "\n",
    "        # Save annotations if not yet set\n",
    "        if variant_data[key]['gene'] == '.':\n",
    "            variant_data[key]['gene'] = record.info.get('GENE_SYMBOL', '.')\n",
    "            variant_data[key]['aa'] = record.info.get('MUTATION_AA', '.')\n",
    "            variant_data[key]['desc'] = record.info.get('MUTATION_DESCRIPTION', '.')\n",
    "\n",
    "# Write output\n",
    "with open(output_file, 'w') as out:\n",
    "    out.write(\"CHROM\\tPOS\\tREF\\tALT\\tCOUNT\\tGENE_SYMBOL\\tMUTATION_AA\\tMUTATION_DESCRIPTION\\n\")\n",
    "    for (chrom, pos, ref, alt), info in variant_data.items():\n",
    "        out.write(f\"{chrom}\\t{pos}\\t{ref}\\t{alt}\\t{info['count']}\\t{info['gene']}\\t{info['aa']}\\t{info['desc']}\\n\")\n",
    "\n",
    "print(\"✅ Done. Output saved to:\", output_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5fcb688d-1c65-4b96-b9a1-bf3e990e1e7c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>COUNT</th>\n",
       "      <th>GENE_SYMBOL</th>\n",
       "      <th>MUTATION_AA</th>\n",
       "      <th>MUTATION_DESCRIPTION</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1664129</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>SKI</td>\n",
       "      <td>p.?</td>\n",
       "      <td>5_prime_UTR_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1664168</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>SKI</td>\n",
       "      <td>p.?</td>\n",
       "      <td>5_prime_UTR_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1664194</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>SKI</td>\n",
       "      <td>p.G7C</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1664208</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>SKI</td>\n",
       "      <td>p.F11=</td>\n",
       "      <td>synonymous_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1664242</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>SKI</td>\n",
       "      <td>p.Q23*</td>\n",
       "      <td>stop_gained</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1081083</th>\n",
       "      <td>chrY</td>\n",
       "      <td>1329683</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>3</td>\n",
       "      <td>P2RY8</td>\n",
       "      <td>p.Y82H</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1081084</th>\n",
       "      <td>chrY</td>\n",
       "      <td>1329712</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "      <td>P2RY8</td>\n",
       "      <td>p.L72P</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1081085</th>\n",
       "      <td>chrY</td>\n",
       "      <td>1329789</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>3</td>\n",
       "      <td>P2RY8</td>\n",
       "      <td>p.W46*</td>\n",
       "      <td>stop_gained</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1081086</th>\n",
       "      <td>chrY</td>\n",
       "      <td>1355143</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>P2RY8</td>\n",
       "      <td>p.?</td>\n",
       "      <td>intron_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1081087</th>\n",
       "      <td>chrY</td>\n",
       "      <td>1394345</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>P2RY8</td>\n",
       "      <td>p.?</td>\n",
       "      <td>intron_variant</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1081088 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        CHROM      POS REF ALT  COUNT GENE_SYMBOL MUTATION_AA  \\\n",
       "0        chr1  1664129   G   T      1         SKI         p.?   \n",
       "1        chr1  1664168   A   C      1         SKI         p.?   \n",
       "2        chr1  1664194   G   T      1         SKI       p.G7C   \n",
       "3        chr1  1664208   C   T      1         SKI      p.F11=   \n",
       "4        chr1  1664242   C   T      1         SKI      p.Q23*   \n",
       "...       ...      ...  ..  ..    ...         ...         ...   \n",
       "1081083  chrY  1329683   A   G      3       P2RY8      p.Y82H   \n",
       "1081084  chrY  1329712   A   G      2       P2RY8      p.L72P   \n",
       "1081085  chrY  1329789   C   T      3       P2RY8      p.W46*   \n",
       "1081086  chrY  1355143   A   C      1       P2RY8         p.?   \n",
       "1081087  chrY  1394345   C   T      1       P2RY8         p.?   \n",
       "\n",
       "        MUTATION_DESCRIPTION  \n",
       "0        5_prime_UTR_variant  \n",
       "1        5_prime_UTR_variant  \n",
       "2           missense_variant  \n",
       "3         synonymous_variant  \n",
       "4                stop_gained  \n",
       "...                      ...  \n",
       "1081083     missense_variant  \n",
       "1081084     missense_variant  \n",
       "1081085          stop_gained  \n",
       "1081086       intron_variant  \n",
       "1081087       intron_variant  \n",
       "\n",
       "[1081088 rows x 8 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "cosmic = pd.read_csv(\"/lustre/scratch126/casm/team274sb/lr26/T2T/cosmic_summaryt2t.tsv\", sep=\"\\t\")\n",
    "cosmic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e2721bd2-bb45-43ef-a9ee-f4f57245e041",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cosmic_filtered = cosmic[cosmic[\"COUNT\"] > 10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ec8ea1f6-b971-4947-8e02-8c8ed2c72c08",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cosmic_filtered.to_csv(\"/lustre/scratch126/casm/team274sb/lr26/T2T/cosmic_summary_hotspots_t2t.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ed434cf9-155e-4f77-a5fc-d1ba24c37973",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of identical rows (ignoring POS column): 323295439\n",
      "Unique rows in file1:\n",
      "                                                       REF  ALT  COUNT  \\\n",
      "19                                                     GCC    G      1   \n",
      "107                                                      A   AA      1   \n",
      "147                                                     TG    T      1   \n",
      "153      TGCTGGGATTACAGGCGTGAGCCACTGCACCTGGCCCCTCCTGACA...    T      1   \n",
      "193                                                     TG    T      2   \n",
      "...                                                    ...  ...    ...   \n",
      "1080719                                               GGCG    G      1   \n",
      "1080724                                                  G  GCG      1   \n",
      "1080739                                                  G   GG      1   \n",
      "1080864                                                  C    A     43   \n",
      "1080965                                                TCT    T      1   \n",
      "\n",
      "        GENE_SYMBOL  MUTATION_AA MUTATION_DESCRIPTION  \n",
      "19              SKI     p.P84del     inframe_deletion  \n",
      "107             SKI          p.?       intron_variant  \n",
      "147             SKI          p.?       intron_variant  \n",
      "153             SKI          p.?       intron_variant  \n",
      "193             SKI          p.?       intron_variant  \n",
      "...             ...          ...                  ...  \n",
      "1080719        FLNA  p.D25Efs*32                  NaN  \n",
      "1080724        FLNA     p.V21dup    inframe_insertion  \n",
      "1080739        FLNA          p.?  5_prime_UTR_variant  \n",
      "1080864       RPL10       p.R98S     missense_variant  \n",
      "1080965       MTCP1          p.?  3_prime_UTR_variant  \n",
      "\n",
      "[47268 rows x 6 columns]\n",
      "Unique rows in file2:\n",
      "                                                       REF ALT  COUNT  \\\n",
      "19                                                     GCC   N      1   \n",
      "107                                                      N  CA      1   \n",
      "151                                                     TG   N      1   \n",
      "157      TGCTGGGATTACAGGCGTGAGCCACTGCACCTGGCCCCTCCTGACA...   N      1   \n",
      "199                                                     TG   N      2   \n",
      "...                                                    ...  ..    ...   \n",
      "1111378                                                  N  GG      1   \n",
      "1111502                                                  C   A     40   \n",
      "1111504                                                  N   A      3   \n",
      "1111602                                                  N   T      1   \n",
      "1111605                                                TCT   N      1   \n",
      "\n",
      "        GENE_SYMBOL  MUTATION_AA MUTATION_DESCRIPTION  \n",
      "19              SKI     p.P84del     inframe_deletion  \n",
      "107             SKI          p.?       intron_variant  \n",
      "151             SKI          p.?       intron_variant  \n",
      "157             SKI          p.?       intron_variant  \n",
      "199             SKI          p.?       intron_variant  \n",
      "...             ...          ...                  ...  \n",
      "1111378        FLNA          p.?  5_prime_UTR_variant  \n",
      "1111502       RPL10       p.R98S     missense_variant  \n",
      "1111504       RPL10  p.R98Tfs*12                  NaN  \n",
      "1111602       MTCP1          p.?  3_prime_UTR_variant  \n",
      "1111605       MTCP1          p.?  3_prime_UTR_variant  \n",
      "\n",
      "[60100 rows x 6 columns]\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load TSV files\n",
    "df1 = pd.read_csv('/lustre/scratch126/casm/team274sb/lr26/T2T/cosmic_summaryt2t.tsv', sep='\\t')\n",
    "df2 = pd.read_csv('/lustre/scratch126/casm/team274sb/lr26/hg38/cosmic_summaryhg38.tsv', sep='\\t')\n",
    "\n",
    "df1\n",
    "df2\n",
    "# Drop the second column (assuming it's always at index 1)\n",
    "df1_dropped = df1.drop(df1.columns[[0,1]], axis=1)\n",
    "df2_dropped = df2.drop(df2.columns[[0,1]], axis=1)\n",
    "\n",
    "# Optional: Reset index for consistent comparison\n",
    "df1_dropped = df1_dropped.reset_index(drop=True)\n",
    "df2_dropped = df2_dropped.reset_index(drop=True)\n",
    "\n",
    "# 1. Count identical rows\n",
    "# Merge and use inner join\n",
    "merged = pd.merge(df1_dropped, df2_dropped, how='inner')\n",
    "identical_count = len(merged)\n",
    "\n",
    "# 2. Find unique rows in each file\n",
    "# Concatenate and drop duplicates that appear twice\n",
    "all_rows = pd.concat([df1_dropped.assign(source='file1'), df2_dropped.assign(source='file2')])\n",
    "unique_rows = all_rows.drop_duplicates(subset=df1_dropped.columns.tolist(), keep=False)\n",
    "\n",
    "# If needed, separate by file\n",
    "unique_file1 = unique_rows[unique_rows['source'] == 'file1'].drop(columns=['source'])\n",
    "unique_file2 = unique_rows[unique_rows['source'] == 'file2'].drop(columns=['source'])\n",
    "\n",
    "# Output results\n",
    "print(\"Number of identical rows (ignoring POS column):\", identical_count)\n",
    "print(\"Unique rows in file1:\")\n",
    "print(unique_file1)\n",
    "print(\"Unique rows in file2:\")\n",
    "print(unique_file2)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "My Py",
   "language": "python",
   "name": "my-python"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
